samza-core/src/main/java/org/apache/samza/operators/impl/WindowOperatorImpl.java

```diff
@@ -18,26 +18,231 @@
+    this.recentCoordinator = coordinator;
+
+    WindowKey<K> storeKey =  getStoreKey(message);
+    BiFunction<M, WV, WV> foldFunction = window.getFoldFunction();
+    WV wv = store.get(storeKey);
+    WV newVal = foldFunction.apply(message, wv);
+    store.put(storeKey, newVal);
+
+    TriggerImpl earlyTriggerImpl = null;
+    TriggerImpl defaultTriggerImpl = null;
+
+    if (earlyTriggers.containsKey(storeKey)) {
+      earlyTriggerImpl = earlyTriggers.get(storeKey);
+      defaultTriggerImpl = defaultTriggers.get(storeKey);
+    } else {
+      // lookup and propagate messages to the right triggers.
+      TriggerContext earlyTriggerContext = new TriggerContextImpl(storeKey);
+      TriggerImpl.TriggerCallbackHandler earlyTriggerHandler = createTriggerHandler(storeKey, TriggerType.EARLY);
+      earlyTriggerImpl = TriggerImpls.createTriggerImpl(window.getEarlyTrigger(), earlyTriggerContext, earlyTriggerHandler);
+
+      TriggerContext defaultTriggerContext = new TriggerContextImpl(storeKey);
+      TriggerImpl.TriggerCallbackHandler defaultTriggerHandler = createTriggerHandler(storeKey, TriggerType.DEFAULT);
+      defaultTriggerImpl = TriggerImpls.createTriggerImpl(window.getDefaultTrigger(), defaultTriggerContext, defaultTriggerHandler);
+
+      earlyTriggers.put(storeKey, earlyTriggerImpl);
+      defaultTriggers.put(storeKey, defaultTriggerImpl);
+    }
+
+    earlyTriggerImpl.onMessage(message);
+    defaultTriggerImpl.onMessage(message);
+  }
+
+  @Override
+  public void onTimer(MessageCollector collector, TaskCoordinator coordinator) {
+    this.recentCollector = collector;
+    this.recentCoordinator = coordinator;
+
+    TriggerTimerState state = pendingCallbacks.peek();
+    if (state == null) {
+      return;
+    }
+
+    long now = System.currentTimeMillis();
+
+    while (state != null && state.getScheduleTimeMs() < now) {
+      pendingCallbacks.remove();
+      state.getCallback().run();
+      state = pendingCallbacks.peek();
+    }
+    super.propagateTimer(collector, coordinator);
+  }
+
+  private WindowKey<K> getStoreKey(M message) {
+    Function<M, K> keyExtractor = window.getKeyExtractor();
+    K key = null;
+
+    if (keyExtractor != null) {
+      key = keyExtractor.apply(message);
+    }
+
+    String windowId = null;
+    if (window.getWindowType() == WindowType.TUMBLING) {
+      long triggerDurationMs = ((TimeTrigger) window.getDefaultTrigger()).getDuration().toMillis();
+      final long now = System.currentTimeMillis();
+      Long windowBoundary = now - now % triggerDurationMs;
+      windowId = windowBoundary.toString();
+    }
+
+    WindowKey<K> windowKey = new WindowKey(key, windowId);
+    return windowKey;
+  }
+
+  private TriggerImpl.TriggerCallbackHandler createTriggerHandler(WindowKey<K> key, TriggerType type) {
+    TriggerImpl.TriggerCallbackHandler handler = new TriggerImpl.TriggerCallbackHandler() {
+      @Override
+      public void onTrigger(TriggerImpl impl, Object key) {
```

Comment:
If we need the most recent collector and coordinator can we make that explicit by taking them as arguments? That would make this much clearer and remove the possibility of not setting the recentCollector / recentCoordinator on a path where it is required.